stdlib/public/runtime/BuiltinProtocolConformances.cpp

```diff
@@ -0,0 +1,805 @@
+            "  .local " SYMBOL(NAME) "\n" \
+            "  .comm " SYMBOL(NAME) ", 128, 16\n" \
+          );
+#elif defined(__MACH__)
+#define CONFORMANCE_PRIVATE_DATA(NAME) \
+          __asm( \
+            "  .zerofill __DATA, __bss, " SYMBOL(NAME) ", 128, 4\n" \
+          );
+#elif defined(_WIN32)
+#define CONFORMANCE_PRIVATE_DATA(NAME) \
+          __asm( \
+            "  .lcomm " SYMBOL(NAME) ", 128, 16\n" \
+          );
+#endif
+
+//===----------------------------------------------------------------------===//
+// Associated Conformances
+//===----------------------------------------------------------------------===//
+
+SWIFT_ASSOCIATED_CONFORMANCE
+_AssociatedConformance _swift_tupleComparable_associatedConformance = {
+  // Header bits for the demangler.
+  255,
+  7,
+
+  // This is a direct relative reference to the base accessor for tuple
+  // Equatable defined below.
+  {
+    0
+  },
+
+  // Null
+  0
+};
+
+//===----------------------------------------------------------------------===//
+// Builtin Conformance Initializer
+//===----------------------------------------------------------------------===//
+
+// Equatable Information
+extern "C" const ProtocolDescriptor EQUATABLE_DESCRIPTOR;
+extern "C" const MethodDescriptor EQUATABLE_EE_METHOD_DESCRIPTOR;
+extern "C" const char __swift_tupleEquatable_private;
+
+// Comparable Information
+extern "C" const ProtocolDescriptor COMPARABLE_DESCRIPTOR;
+extern "C" const MethodDescriptor COMPARABLE_BASE_CONFORMANCE_DESCRIPTOR;
+extern "C" const MethodDescriptor COMPARABLE_LT_METHOD_DESCRIPTOR;
+extern "C" const MethodDescriptor COMPARBALE_LTE_METHOD_DESCRIPTOR;
+extern "C" const MethodDescriptor COMPARABLE_GTE_METHOD_DESCRIPTOR;
+extern "C" const MethodDescriptor COMPARABLE_GT_METHOD_DESCRIPTOR;
+extern "C" const char __swift_tupleComparable_private;
+
+// Hashable Information
+extern "C" const ProtocolDescriptor HASHABLE_DESCRIPTOR;
+extern "C" const MethodDescriptor HASHABLE_BASE_CONFORMANCE_DESCRIPTOR;
+extern "C" const MethodDescriptor HASHABLE_HASHVALUE_METHOD_DESCRIPTOR;
+extern "C" const MethodDescriptor HASHABLE_HASH_METHOD_DESCRIPTOR;
+extern "C" const MethodDescriptor HASHABLE_RAWHASHVALUE_METHOD_DESCRIPTOR;
+extern "C" const char __swift_tupleHashable_private;
+
+// This is needed to properly emplace relative references within the data
+// structures defined in this file. Because we can't directly express relative
+// references within C++, we have to calculate the offsets ourselves. A
+// previous version used assembly to work around this, but the use of assembly
+// is very non-portable and makes it incredibly easy to get things wrong.
+__attribute__((__constructor__))
+static void initializeBuiltinConformances() {
+  //---------------------------------//
+  // Tuple Equatable Initialization
+  //---------------------------------//
+
+  // Setup the protocol descriptor.
+  _swift_tupleEquatable_conf.protocol.set(&EQUATABLE_DESCRIPTOR);
```

Comment:
Argh, we're trying really hard not to have any static initializers in the runtime. I don't even know if it's possible to express in C(++) the initialization of relative references like this, since we're essentially trying to trick Clang into emitting the kind of IR that IRGen hand-crafts to get the effect we want.

The fact that everything is being manually wired up like this anyway makes me wonder if we should put the hard-coded logic into IRGen itself. For example, what we if created an instance or subclass of `ProtocolConformanceDescriptorBuilder` and `WitnessTableBuilder` and feed it the bespoke witnesses and metadata we need, so we can rely on IRGen's knowledge of how to emit relative references, the structure of descriptors / witness tables, and so on.

Reply:
I was wondering if we could do something better as well. I wonder if we could instead have an approach that leverages Swift/IRGen in a nice way. Specifically:

What if we made it so that there was a builtin type that referred to Builtins that work for all Tuples regardless of length. Then, we could write the conformance part of this as a conformance for that dummy type and then just call out to the runtime in the conformance to the c++ code with space for marshaling.